src/invert/laplace/impls/iterative_parallel_tri/iterative_parallel_tri.cxx

```diff
@@ -0,0 +1,1597 @@
+      subcount=0;
+    }
+
+    // Throw error if we are performing too many iterations
+    if (count>maxits) {
+      // Maximum number of allowed iterations reached.
+      // If the coarsest multigrid iteration matrix is diagonally-dominant,
+      // then convergence is guaranteed, so maxits is set too low.
+      // Otherwise, the method may or may not converge.
+      bool is_dd = true;
+      for (int kz = 0; kz < nmode; kz++) {
+	is_dd = is_diagonally_dominant(levels[max_level],jy,kz);
+	if(not is_dd){
+	  break;
+	}
+      }
+
+      bool global_is_dd;
+      MPI_Allreduce(&is_dd, &global_is_dd, 1, MPI::BOOL, MPI_LAND, BoutComm::get());
+
+      if(global_is_dd){
+	throw BoutException("LaplaceIPT error: Not converged within maxits=%i iterations. The coarsest grained iteration matrix is diagonally dominant and convergence is guaranteed. Please increase maxits and retry.",maxits);
+      }
+      else{
+	throw BoutException("LaplaceIPT error: Not converged within maxits=%i iterations. The coarsest iteration matrix is not diagonally dominant so there is no guarantee this method will converge. Consider (1) increasing maxits; or (2) increasing the number of levels (as grids become more diagonally dominant with coarsening). Using more grids may require larger NXPE.",maxits);
+      }
+    }
+  }
+  ///SCOREP_USER_REGION_END(whileloop);
+  ///SCOREP_USER_REGION_DEFINE(afterloop);
+  ///SCOREP_USER_REGION_BEGIN(afterloop, "after faff",///SCOREP_USER_REGION_TYPE_COMMON);
+
+#if CHECK > 2
+  for(int ix=0;ix<4;ix++){
+    for(int kz=0;kz<nmode;kz++){
+      if(!finite(levels[0].xloc(ix,kz).real()) or !finite(levels[0].xloc(ix,kz).imag()))
+	throw BoutException("Non-finite xloc at %d, %d, %d", ix, jy, kz);
+    }
+  }
+#endif
+
+  // Cache solution
+  for(int ix=0; ix<4; ix++){
+    for(int kz=0; kz<nmode; kz++){
+      x0saved(jy,ix,kz) = levels[0].xloc(ix,kz);
+    }
+  }
+
+  reconstruct_full_solution(xk1d,levels[0],jy);
+
+#if CHECK > 2
+  for(int ix=0;ix<ncx;ix++){
+    for(int kz=0;kz<nmode;kz++){
+      if(!finite(xk1d(kz,ix).real()) or !finite(xk1d(kz,ix).imag()))
+	throw BoutException("Non-finite xloc at %d, %d, %d", ix, jy, kz);
+    }
+  }
+#endif
+
+  ++ncalls;
+  ipt_mean_its = (ipt_mean_its * BoutReal(ncalls-1)
+  + BoutReal(count))/BoutReal(ncalls);
+
+  for (int kz = 0; kz <= maxmode; kz++) {
+    // If the global flag is set to INVERT_KX_ZERO
+    if ((global_flags & INVERT_KX_ZERO) && (kz == 0)) {
+      dcomplex offset(0.0);
+      for (int ix = localmesh->xstart; ix <= localmesh->xend; ix++) {
+	offset += xk1d(kz,ix);
+      }
+      offset /= static_cast<BoutReal>(localmesh->xend - localmesh->xstart + 1);
+      for (int ix = localmesh->xstart; ix <= localmesh->xend; ix++) {
+	xk1d(kz,ix) -= offset;
+      }
+    }
+
+    // Store the solution xk for the current fourier mode in a 2D array
+    for (int ix = 0; ix < ncx; ix++) {
+      xk(ix, kz) = xk1d(kz,ix);
+    }
```

Comment:
It's because FFTs need `kz` fastest moving while Lapack needs `ix` fastest moving.  Easier to just make a transpose function I think.  

At the moment this is profiled in the same code section as the Thomas algorithm solve.  Would be interesting to separate the two out.

Reply:
I thought you probably had a good reason for having the indices like this :smile: